From 42f3737177d6269bbb3e826f8357dc6559e35a14 Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Thu, 31 Jul 2014 02:51:51 -0400
Subject: [drm] qxl: return IRQ_NONE if it was not our irq

Message-id: <1406775111-19208-1-git-send-email-jasowang@redhat.com>
Patchwork-id: 86855
O-Subject: [RHEL7.1 PATCH] drm/qxl: return IRQ_NONE if it was not our irq
Bugzilla: 1095099
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>

Bugzilla: 1095099
Brew Build: https://brewweb.devel.redhat.com/taskinfo?taskID=7764225
Test status: Tested by QE

Return IRQ_NONE if it was not our irq. This is necessary for the case
when qxl is sharing irq line with a device A in a crash kernel. If qxl
is initialized before A and A's irq was raised during this gap,
returning IRQ_HANDLED in this case will cause this irq to be raised
again after EOI since kernel think it was handled but in fact it was
not.

Cc: Gerd Hoffmann <kraxel@redhat.com>
Cc: stable@vger.kernel.org
Signed-off-by: Jason Wang <jasowang@redhat.com>
Signed-off-by: Dave Airlie <airlied@redhat.com>
(cherry picked from commit fbb60fe35ad579b511de8604b06a30b43846473b)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/qxl/qxl_irq.c b/drivers/gpu/drm/qxl/qxl_irq.c
index 21393dc..f4b6b89 100644
--- a/drivers/gpu/drm/qxl/qxl_irq.c
+++ b/drivers/gpu/drm/qxl/qxl_irq.c
@@ -33,6 +33,9 @@ irqreturn_t qxl_irq_handler(DRM_IRQ_ARGS)
 
  pending = xchg(&qdev->ram_header->int_pending, 0);
 
+ if (!pending)
+  return IRQ_NONE;
+
  atomic_inc(&qdev->irq_received);
 
  if (pending & QXL_INTERRUPT_DISPLAY) {
-- 
1.7.1